packages/core-data/src/types/README.md

```diff
@@ -0,0 +1,211 @@
+# Entity Records Types
+
+!THIS IS WORK IN PROGRESS! I used bits of the GitHub discussion as the first draft here
+
+## Overview
+
+The types in this directory are designed to support the following use-cases:
+
+* Context-aware type checks for entity records.
+* Static type checks for *edited* entity records, where certain fields become strings instead of objects.
```

Comment:
This is somewhat misleading. As I've tried to be obnoxious about, these do not provide type safety for API responses. More or less they help us provide auto-complete when working with API data and yes, they add type checks but only for the `Updatable` fields.

Maybe something could be reworded to acknowledge that without making people feel like these are useless.

```md
* Provide type-hinting and documentation for entity records fetched in the various contexts.
* Type-check the values we use to *edit* entity records, the values that are sent back to the server as updates.
```

Not a fan of my own writing here. Maybe the entire overview could be different. Don't let me be too pessimistic here; I just really want to scare people away from the idea that you can have safety when injecting outside data sources into an application without parsing. The reality is that people will end up with type errors in their code despite these definitions because we don't have as a practice any step that ensures this. They shouldn't be surprised when the value they get from the server doesn't match the types.

Reply:
How about a warning like below?

---

* Provide type-hinting and documentation for entity records fetched in the various REST API contexts.
* Type-check the values we use to *edit* entity records, the values that are sent back to the server as updates.

**Warning:** The types model the _expected_ API responses which is **not** the same as having a full type safety for API-related operations. The API responses are **not** parsed at the moment and there are many cases where they could disagree with the types defined here, like bugs in the code, a newer API version, or a plugin modifying the response.

Reply:
I like that warning, though I think the API responses _are_ used as-is and that is the cause of the problem. Was that emphatic "not" supposed to be there? If there's nuance then rewording could avoid it: "The API responses are type-cast to these definitions and therefore may not match those expectations; for example, …"

Reply:
Thanks! I adjusted the wording in [466dd5f](https://github.com/WordPress/gutenberg/pull/38666/commits/466dd5f360d202a7bfb3f1676ff7b9ef0239d876).